Add MachineTranslatedText storage model and domain object

SarahZLipeles opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #N/A.
2. This PR does the following: Creates the MachineTranslatedText storage model and domain object for the [Computer Aided Translation project](https://docs.google.com/document/d/1kJd-yLTzB9a2c3Nq7v9pzKfHwKHKGpkWfQ8B0YGf50U/edit?usp=sharing).

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes made in this PR work correctly.
If the changes in your PRs are autogenerated via a script and you cannot provide proof for the changes then please leave a comment "No proof of changes needed because {{Reason}}".
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

Commit:
workspace transfer

Commit:
formatting

Commit:
move MachineTranslatedText to translation_domain.py

Commit:
import utils convert to hash

Commit:
Merge remote-tracking branch 'upstream/develop' into machine-translated-text

Commit:
add storage model

Commit:
add translation domain object and tests

Commit:
Merge remote-tracking branch 'upstream/develop' into machine-translated-text

Commit:
add codeowners

Commit:
fix typo

seanlip commented on core/domain/translation_domain.py:
```diff
@@ -0,0 +1,109 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Models for machine translations retrieved from Google Cloud Translate."""
```
Add a line above this (I think we do this in other files).

SarahZLipeles replied:
Done. I think "Models" is also an inaccurate descriptor here, I changed it to "domain objects".

Commit:
change code owner to @vojtechjelinek

seanlip commented on core/domain/translation_domain.py:
```diff
@@ -0,0 +1,109 @@
+                'Expected source_language_code to be a string, received %s' %
+                self.source_language_code
+            )
+        if not utils.is_supported_audio_language_code(
```
Could you please add `TODO(#12341): Tidy up this logic once we have a canonical list of language codes.` here and below, so that @DubeySandeep sees it? (He's working on #12341.)

SarahZLipeles replied:
Added.

seanlip commented on core/storage/translation/gae_models.py:
```diff
@@ -0,0 +1,137 @@
+
+
+class MachineTranslatedTextModel(base_models.BaseModel):
+    """Model for storing machine translations"""
```
Missing period at end.

Also, explain in the docstring how the ID is constructed.

SarahZLipeles replied:
Done.

seanlip commented on core/storage/translation/gae_models.py:
```diff
@@ -0,0 +1,137 @@
+    """Model for storing machine translations"""
+
+    # The untranslated text.
+    origin_text = datastore_services.StringProperty(required=True)
```
I would expect origin_text and translated_text to either both be StringProperty or both be TextProperty. Also, the indexing requirements should be the same for both. Could you explain the discrepancy?

Note that StringProperty is for strings < 1500 chars and TextProperty is for strings greater than that. The latter cannot be indexed.

(I suspect you'll go with TextProperty. In that case, though, perhaps adding an extra StringProperty indexed field for the hashed_origin_text would be sensible.)

SarahZLipeles replied:
This was a vestigial mistake, I changed it to TextProperty with indexed=False.
Do you think the hashed_origin_text is needed? What use cases were you thinking of?

seanlip replied:
Yup, I think it would be useful. In the future, we might want to query for all translations for a particular string (in different languages); having such a field would let us hash that string and retrieve all translations for it.

SarahZLipeles replied:
Makes sense! Fixed.

seanlip commented on core/storage/translation/gae_models.py:
```diff
@@ -0,0 +1,137 @@
+    @classmethod
+    def _generate_id(
+            cls, source_language_code, target_language_code, origin_text):
+        """Generates key for the instance of MachineTranslatedTextModel
```
key --> a key

SarahZLipeles replied:
Fixed.

seanlip commented on core/storage/translation/gae_models.py:
```diff
@@ -0,0 +1,137 @@
+        """
+        # max_length is 450, less than maximum key length (500) to allow room
+        # for language_codes.
+        hashed_text = utils.convert_to_hash(origin_text, 450)
```
Does utils.convert_to_hash() ever return something that's more than 450 characters? If not then it might make sense to use a smaller max length that's more "in range".

SarahZLipeles replied:
I think utils.convert_to_hash() doesn't return anything over 50 characters. I changed it accordingly.

SarahZLipeles commented:
NOTE: This PR is missing the validators, I'm still working on them.

Commit:
add translation_validators

Commit:
sort imports
fix spacing

Commit:
fix origin_text type
Change max hash length to be more "in range"
Fix/clarify docstrings and comments
Add TODOs for tidying language code verification logic related to #12341

Commit:
fix spacing

Commit:
add MachineTranslatedTextModelAuditOneOffJob to job registry

Commit:
fix typo

SarahZLipeles commented:
@seanlip I've addressed your comments and added the validators, PTAL.

SarahZLipeles commented:
@sagangwee PTAL

seanlip commented on core/domain/translation_domain.py:
```diff
@@ -13,7 +13,9 @@
 # limitations under the License.
-"""Models for machine translations retrieved from Google Cloud Translate."""
+
+"""Domain objects for machine translations retrieved from Google Cloud
```
`Domain objects related to translations.`

(We might put other stuff in this file in future.)

SarahZLipeles replied:
Fixed.

seanlip commented:
See also https://github.com/oppia/oppia/pull/12515#discussion_r613648386

seanlip commented:
@TheoLipeles PTAL, left two more comments.

Also adding @BenHenning @brianrodri @DubeySandeep @U8NWXD @vojtechjelinek as codeowner reviewers -- PTAL, thanks!

sagangwee commented on core/domain/translation_domain.py:
```diff
@@ -0,0 +1,115 @@
+            ValidationError. One or more attributes of the MachineTranslatedText
+                are invalid.
+        """
+        # validate source_language_code.
```
Don't think this comment is really necessary as it's pretty clear what the validation is. Here and in the other validations.

SarahZLipeles replied:
These comments were leftover from my comment scaffolding. I removed them.

sagangwee commented on core/domain/translation_domain.py:
```diff
@@ -0,0 +1,115 @@
+            )
+
+    def to_dict(self):
+        """Convert the MachineTranslatedText domain instance into a dictionary
```
Nit: Prefer to write in third-person for verb phrases to be consistent, i.e. "Converts..."

SarahZLipeles replied:
Fixed.

sagangwee commented on core/domain/translation_domain_test.py:
```diff
@@ -0,0 +1,107 @@
+
+
+class MachineTranslatedTextTests(test_utils.GenericTestBase):
+    """Tests for the MachineTranslatedTexts domain object."""
```
Nit: `MachineTranslatedTexts` should be singular to match the class name, right? Here and below.

SarahZLipeles replied:
Thanks for catching this! Fixed.

sagangwee commented on core/domain/translation_domain_test.py:
```diff
@@ -0,0 +1,107 @@
+        self.translation.validate()
+
+    def test_validate_with_non_string_source_language_code_raises(self):
+        self._init_translation()
```
Aren't these initialize lines redundant in the tests as it's being called in the setup?

SarahZLipeles replied:
I don't fully understand how the setup function is used, but it doesn't appear to run before each test (I ran without these lines and got self.translation is NoneType).

DubeySandeep replied:
It should be `def setUp`, can you please check whether it works with setUp?

SarahZLipeles replied:
After changing it to `setUp` this works without the redundant _inits.
Fixed.

sagangwee commented on core/domain/translation_validators.py:
```diff
@@ -0,0 +1,54 @@
+
+class MachineTranslatedTextModelValidator(
+        base_model_validators.BaseModelValidator):
+    """Class for validating MachineTranslationModel."""
```
Nit: You can omit "Class" here and just use a verb phrase, i.e. "Validates MachineTranslationModel."

SarahZLipeles replied:
Fixed.

sagangwee commented on core/domain/translation_validators.py:
```diff
@@ -0,0 +1,54 @@
+        language_codes_regex = '(%s)' % '|'.join(
+            cloud_translate_services.LANGUAGE_CODE_ALLOWLIST)
+        # Valid id: [source_language_code]:[target_language_code]:[hashed_text].
+        regex_string = '^%s\\:%s\\:.+$' % (
```
Are there backslashes in the ID? Wouldn't this match backslashes before the colons? Do the validation checks fail if you omit the backslashes?

SarahZLipeles replied:
Ah, thanks for catching this, I was originally using `|` as the separation character, the validations run fine without it. Fixed.

sagangwee commented on core/storage/translation/gae_models.py:
```diff
@@ -0,0 +1,141 @@
+    """Model for storing machine translations. Model instances have a key
+    generated from the source and target language codes, followed by a sha-1
+    hash of the origin text formated as follows:
+        '[source_language_code]:[target_language_code]:[hashed_text]'
```
Nit: Can we add a new line before and after this line?

SarahZLipeles replied:
Added. Also added in `_generate_id` below.

sagangwee commented on core/storage/translation/gae_models.py:
```diff
@@ -0,0 +1,141 @@
+            translated_text: str. The translation.
+
+        Returns:
+            MachineTranslatedTextModel. The newly created
```
Doesn't this return the ID?

SarahZLipeles replied:
Yup. Fixed.

sagangwee commented on core/storage/translation/gae_models.py:
```diff
@@ -0,0 +1,141 @@
+            MachineTranslatedTextModel. The newly created
+            MachineTranslatedTextModel instance.
+        """
+
```
Remove this line?

SarahZLipeles replied:
Fixed.

sagangwee commented on core/storage/translation/gae_models.py:
```diff
@@ -0,0 +1,141 @@
+        """
+        # max_length is 450, less than maximum key length (500) to allow room
+        # for language_codes.
+        hashed_text = utils.convert_to_hash(origin_text, 50)
```
Should the max length be set to 450 like the comment says? Also nit: prefer to capitalize the comment sentence, especially if you're adding a period to make it a "real" sentence".

SarahZLipeles replied:
I forgot to remove this comment after addressing [Sean's comment](https://github.com/oppia/oppia/pull/12515#discussion_r613610177).

sagangwee commented on core/storage/translation/gae_models.py:
```diff
@@ -0,0 +1,141 @@
+    @classmethod
+    def _generate_id(
+            cls, source_language_code, target_language_code, origin_text):
+        """Generates a key for the instance of MachineTranslatedTextModel
+        class in the required format with the arguments provided.
```
I think this can be simplified to: "Generates a valid key for a MachineTranslatedTextModel."

SarahZLipeles replied:
Thanks! Done.

sagangwee commented on core/storage/translation/gae_models.py:
```diff
@@ -0,0 +1,141 @@
+
+    @staticmethod
+    def get_model_association_to_user():
+        """Model does not contain user data."""
```
Nit: Use "doesn't" to be consistent with above and below.

SarahZLipeles replied:
Done. Is it better to prefer does not to avoid the abbreviation?

sagangwee commented on core/storage/translation/gae_models_test.py:
```diff
@@ -0,0 +1,75 @@
+
+class MachineTranslatedTextModelTests(test_utils.GenericTestBase):
+    def test_create_model(self):
+
```
Remove this line?

TheoLipeles replied:
Fixed.

SarahZLipeles replied:
Fixed.

sagangwee reviewed (changes requested):
@TheoLipeles Took a pass, thanks!

Commit:
add hashed_origin_text field to MachineTranslatedTextModel
(allows lookup by origin_text)

Commit:
address pr comments:
change MachineTranslatedTextTests name to be consistent
clarify/clean-up comments
remove unnecessary backslashes from regex
fix spacing

SarahZLipeles commented:
@seanlip @sagangwee I addressed your comments, PTAL.

seanlip reviewed (approved):
LGTM from my end. Thanks!

Commit:
Merge remote-tracking branch 'upstream/develop' into machine-translated-text

Commit:
fix typo

Commit:
add translation fetchers

SarahZLipeles commented:
Note to reviewers: This PR is somewhat urgent, It's blocking my next two milestones which I'm aiming to complete by the end of Friday.

DubeySandeep commented on .github/CODEOWNERS:
```diff
@@ -113,6 +113,8 @@
 /core/templates/domain/opportunity/ @sagangwee
 /core/templates/pages/contributor-dashboard-page/ @sagangwee
+/core/domain/translation*.py @vojtechjelinek
+/core/storage/translation/ @vojtechjelinek
```
Add items here in sorted order*

SarahZLipeles replied:
Fixed.

vojtechjelinek commented on core/domain/prod_validation_jobs_one_off.py:
```diff
@@ -769,6 +769,15 @@ def entity_classes_to_map_over(cls):
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [
+            translation_models.MachineTranslatedTextModel]
```
```suggestion
        return [translation_models.MachineTranslatedTextModel]
```

SarahZLipeles replied:
Fixed.

vojtechjelinek commented on core/storage/translation/gae_models.py:
```diff
@@ -0,0 +1,147 @@
+    generated from the source and target language codes, followed by a sha-1
+    hash of the origin text formated as follows:
+
+        [source_language_code]:[target_language_code]:[hashed_origin_text]
```
In many other places we use either `.` or `-` to divide different parts of the ID, why is `:` used here?

SarahZLipeles replied:
I originally was using the `|` character but replaced it with `:` to make the model id regex more readable. I prefer to avoid using `-` because it sometimes appears in language codes, and `.` is a special character in regular expressions as well.
Other than readability though, there's no reason not to use `.`. Do you suggest changing it?

vojtechjelinek replied:
Yep, lets use `.` since it used accross the codebase and you can always escape it in the regexes.

SarahZLipeles replied:
Fixed.

vojtechjelinek reviewed (changes requested):
Thanks! Left a two comments.

Commit:
fix spacing

SarahZLipeles commented:
@vojtechjelinek I addressed your comments, PTAL.

DubeySandeep commented on core/domain/translation_domain_test.py:
```diff
@@ -0,0 +1,107 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests for translate services."""
```
... translation domain objects*

SarahZLipeles replied:
Fixed.

DubeySandeep commented on core/domain/translation_domain_test.py:
```diff
@@ -0,0 +1,107 @@
+
+    translation = None
+
+    def setup(self):
```
set**U**p*

SarahZLipeles replied:
Thank you! I've been wracking my brain over that one. Fixed.

DubeySandeep reviewed (changes requested):
LGTM, left a few minor comments PTAL!

Commit:
fix misnamed setup
sort codeowners lines

SarahZLipeles commented:
@DubeySandeep Thanks for the quick response! I addressed your comments, PTAL.

sagangwee commented on core/storage/translation/gae_models.py:
```diff
@@ -0,0 +1,147 @@
+
+class MachineTranslatedTextModel(base_models.BaseModel):
+    """Model for storing machine translations. Model instances have a key
+    generated from the source and target language codes, followed by a sha-1
```
Nit: capitalize "sha-1". Here and below. Also, try to be consistent with the article. I've heard SHA as both an initialism and an acronym (i.e. like "shaw"), so I'd say choose one and stick to it (though I believe its use as an acronym is more common).

SarahZLipeles replied:
Fixed. I went with SHA-1.

BenHenning commented on core/storage/translation/gae_models.py:
```diff
@@ -0,0 +1,147 @@
+    hashed_origin_text = datastore_services.StringProperty(
+        required=True, indexed=True)
+    # The language code of the untranslated text.
+    source_language_code = datastore_services.StringProperty(
```
We refer to the untranslated text above as 'origin' but 'source' here. Can we consolidate on a single term?

Also, maybe 'original' would be clearer. For example: original_text_language_code.

SarahZLipeles replied:
I consolidated to "source" and "target" to match the Google Cloud Translate API.

sagangwee commented on core/storage/translation/gae_models.py:
```diff
@@ -0,0 +1,147 @@
+
+            [source_language_code]:[target_language_code]:[hashed_origin_text].
+        """
+
```
Remove this line?

SarahZLipeles replied:
Done.

BenHenning commented on core/storage/translation/gae_models.py:
```diff
@@ -0,0 +1,147 @@
+
+    # The untranslated text.
+    origin_text = datastore_services.TextProperty(required=True, indexed=False)
+    # An sha-1 hash of the origin text.
```
I suggest adding some context here about how this is intended to be used (not necessarily how it is, but what the intention is for storing it in the storage model).

SarahZLipeles replied:
Added.

BenHenning commented on core/storage/translation/gae_models.py:
```diff
@@ -0,0 +1,147 @@
+    source_language_code = datastore_services.StringProperty(
+        required=True, indexed=True)
+    # The language code of the translation.
+    target_language_code = datastore_services.StringProperty(
```
Similarly here: we should use the same term for the translated result. Perhaps: translated_text_language_code?

Also a nit: we order language code below text above--I suggest matching that here for organizational parity.

SarahZLipeles replied:
I consolidated to "source" and "target" to match the Google Cloud Translate API.

BenHenning replied:
Hmm I was expecting the language code to match ``translated_text`` as well (e.g. ``target_text``), though I feel less strongly about this than the source/origin text. Deferring to you whether you want to make these two consistent.

SarahZLipeles replied:
I'm sticking with `translated_text` here, I read "target" as referring to the target language rather than the end translation.

BenHenning commented on core/storage/translation/gae_models.py:
```diff
@@ -0,0 +1,147 @@
+        Returns:
+            str. The id of the newly created MachineTranslatedTextInstance.
+        """
+        hashed_origin_text = utils.convert_to_hash(origin_text, 50)
```
Not related to  your PR, but @seanlip why do we have a length limit for convert_to_hash? Unless I'm mistaken, SHA-1 always produces a 160 bit string.

Further, we might want to consider being explicit about the hash name. If storage models & other APIs specifically need to depend on SHA-1, I'd expect them to call something like ``utils.convert_to_sha1_hash`` or similar to ensure their contracts can be upheld.

SarahZLipeles replied:
I added a note here clarifying the choice of max_length.

sagangwee commented on core/storage/translation/gae_models.py:
```diff
@@ -0,0 +1,147 @@
+        translation_entity.put()
+        return entity_id
+
+    @classmethod
```
Not exactly sure why we need this decorator here. I see this being used in other `_generate_id` methods in the codebase, but it doesn't really seem too add much value especially since it's a "private" method anyway.

SarahZLipeles replied:
I changed this to be a static method since it's agnostic to model instances.

BenHenning commented on core/storage/translation/gae_models.py:
```diff
@@ -0,0 +1,147 @@
+    generated from the source and target language codes, followed by a sha-1
+    hash of the origin text formated as follows:
+
+        [source_language_code]:[target_language_code]:[hashed_origin_text]
```
Is the intention with this ID mapping to reuse translations if the text is the same? That might be fine for machine translations, but for manual translations we should also factor in the context since translations can change for different contexts even when the original English text is the same.

SarahZLipeles replied:
Yes, this storage model exists to prevent hitting the Cloud Translate API more than once for the same translation. It is only relevant for machine translations which are displayed as a helper in the contributor dashboard translation opportunity modal. See the [Computer Aided Translation project](https://docs.google.com/document/d/1kJd-yLTzB9a2c3Nq7v9pzKfHwKHKGpkWfQ8B0YGf50U/edit?usp=sharing) for more details.

BenHenning replied:
Hmm, I think it might be worth indicating in the model's docstring that models are expected to be mapped this way (e.g. that it's okay we're using the same translation irrespective of context because of relevant reasons, maybe: these translations are only meant to be used as a reference, and we want to avoid hitting the translation API multiple times since it will presumably provide consistent translations for the same input string).

SarahZLipeles replied:
I added a little more detail here. Does this read more clearly, or is it too verbose?

BenHenning commented on core/storage/translation/gae_models.py:
```diff
@@ -0,0 +1,147 @@
+        Args:
+            source_language_code: str. The language of the origin_text.
+            target_language_code: str. The language being translated to.
+            origin_text: str. The text to be translated.
```
Slight nit, but maybe instead use: ``The text to be or that was translated.`` (since in this context we may be fetching something already translated).

BenHenning replied:
Though, this actually begs a different question: will we always know the translated text upon creation? If so, then perhaps ``The text that was, or may need to be, translated.`` would actually be clearer.

SarahZLipeles replied:
I updated the language in the docstrings to be more descriptive, PTAL.

BenHenning commented on core/storage/translation/gae_models.py:
```diff
@@ -0,0 +1,147 @@
+
+    @staticmethod
+    def get_model_association_to_user():
+        """Model doesn't contain user data."""
```
```suggestion
        """Model is not associated with users."""
```

I think this might be a bit clearer.

SarahZLipeles replied:
Fixed.

BenHenning commented on core/storage/translation/gae_models_test.py:
```diff
@@ -0,0 +1,75 @@
+    [models.NAMES.base_model, models.NAMES.translation])
+
+
+class MachineTranslatedTextModelTests(test_utils.GenericTestBase):
```
Other test cases that I suggest including:
- Retrieving a model that doesn't yet exist
- get_translation_for_text for a model that doesn't exist
- get_translation_for_text for a model that does exist
- Optional: get_translation_for_text for a model that does exist, but with incorrect parameters (e.g. slightly wrong original text, e.g. if it was updated elsewhere in the system)
- Optional: using a non-English source code
- Optional: two different translations for the same source/target language codes & verifying that they're different (to test entity key uniqueness)

SarahZLipeles replied:
I added a variety of tests here, let me know if you think there are more cases to cover.

BenHenning replied:
Current test coverage seems reasonable--thanks.

BenHenning reviewed (commented):
Thanks @TheoLipeles! Apologies for the delayed review. Had mostly nit comments for the storage layer--PTAL.

sagangwee reviewed (approved):
@TheoLipeles LGTM, minus a few nits. Thanks!

vojtechjelinek reviewed (approved):
LGTM!

Commit:
fix formating and docstring nits

Commit:
change _generate_id to be a static method

Commit:
add tests for get_translation_for_text
clarify docstrings

Commit:
consolidate terminology:
    replace "origin" with "source" in reference to the untranslated source text

SarahZLipeles commented:
@BenHenning I addressed your comments and renovated the docstrings in `gae_models.py`, PTAL.

seanlip commented:
Just a note: @brianrodri is away so I'll cover his review.

@BenHenning @DubeySandeep PTAL; I can merge this once you're happy with it. Thanks!

BenHenning commented on core/storage/translation/gae_models.py:
```diff
@@ -29,118 +29,137 @@
-    hashed_origin_text = datastore_services.StringProperty(
+    # The untranslated source text.
+    source_text = datastore_services.TextProperty(required=True, indexed=False)
+    # An SHA-1 hash of the source text. This can be used to index the datastore
```
Nit: ``A SHA-1`` for consistency with your description above (and I *think* phonetically pronouncing 'SHA-1' is more conventional, but I couldn't find consensus on this when searching).

SarahZLipeles replied:
Fixed.

TheoLipeles replied:
Fixed.

BenHenning reviewed (approved):
Thanks @TheoLipeles! Just had a couple more comments. Approving since the PR LGTM per codeowners, but let me know if you want me to check any follow-up documentation changes.

Commit:
add more description regarding key purpose

seanlip commented:
@TheoLipeles -- just flagging this too: https://github.com/oppia/oppia/pull/12515#discussion_r614384547

Commit:
replace ":" with "." as a divider character to conform to codebase practices

DubeySandeep commented on .github/CODEOWNERS:
```diff
@@ -110,7 +110,9 @@
 # Contributor Dashboard project.
 /core/controllers/contributor_dashboard*.py @sagangwee
 /core/domain/opportunity*.py @sagangwee
+/core/domain/translation*.py @vojtechjelinek
```
Why is this assigned to @vojtechjelinek are we expecting it to be assigned to @sagangwee (Contributor dashboard team)?

SarahZLipeles replied:
This was suggested by @seanlip.

seanlip replied:
Yup, @vojtechjelinek is correct, I already checked with him. He is working on a project to migrate translations to their own models, so this fits in with that too.

DubeySandeep reviewed (approved):
LGTM, left one comment, PTAL!

seanlip commented:
Merging, since @brianrodri is away. Thanks @TheoLipeles!
